packages/xforms-engine/test/parse/xpath/dependency-analysis.test.ts

```diff
@@ -0,0 +1,337 @@
+
+			contextNodeset: 'foo',
+			expression: '/data/bar[bat = /data/quux] | quux[./zig = ../zag or /data/beep = false()]',
+			expected: [
+				'/data/bar',
+				'/data/bar/bat',
+				'/data/quux',
+				'foo/quux',
+				'foo/quux/zig',
+				'foo/zag',
+				'/data/beep',
+			],
+		},
+
+		{
+			description:
+				'Absolute expression context; absolute expression -> stripped predicate; relative predicate expression resolved to that; predicate references beginning with current() FilterExpr -> relative to expression context',
+
+			contextNodeset: '/data/foo',
+			expression: '/data/bar[./quux or current()/bat or current()/../quux]',
+			expected: ['/data/bar', '/data/bar/quux', '/data/foo/bat', '/data/quux'],
+		},
+
+		{
+			description: 'Multiple self steps elided',
+
+			contextNodeset: '/data/foo',
+			expression: './././bar/./quux',
+			expected: ['/data/foo/bar/quux'],
+		},
+
+		{
+			description: 'Multiple parent steps resolved',
+
+			contextNodeset: '/data/foo/bar/bat/quux/zig/zag',
+			expression:
+				'concat(../sibling, ../../parent-sibling, ../../../impertinent/../grandparent-sibling)',
+			expected: [
+				'/data/foo/bar/bat/quux/zig/sibling',
+				'/data/foo/bar/bat/quux/parent-sibling',
+				'/data/foo/bar/bat/grandparent-sibling',
+			],
+		},
+
+		{
+			description: 'Absolute context; self reference (direct or redundant) resolved to context',
+
+			contextNodeset: '/data/foo',
+			expression: '. = 1 or ./. = 2',
+			expected: ['/data/foo'],
+		},
+
+		{
+			description: 'Relative context; self reference (direct or redundant) resolved to context',
+
+			contextNodeset: 'foo',
+			expression: '. != false() and ./. != false()',
+			expected: ['foo'],
+		},
+
+		{
+			description: 'No context; self reference preserved',
+
+			contextNodeset: null,
+			expression: '. > now()',
+			expected: ['.'],
+		},
+
+		{
+			description: 'No context; redundant self reference simplified',
+
+			contextNodeset: null,
+			expression: './. > now()',
+			expected: ['.'],
+		},
+
+		{
+			description: 'Unabbreviated self::node() and self::* treated as .',
+
+			contextNodeset: '/data/foo',
+			expression: 'self::node()/bar | self::node()/quux',
+			expected: ['/data/foo/bar', '/data/foo/quux'],
+		},
+
+		{
+			description: 'Unabbreviated parent::node() and parent::* treated as ..',
+
+			contextNodeset: '/data/foo',
+			expression: 'parent::node()/bar and parent::*/quux',
+			expected: ['/data/bar', '/data/quux'],
+		},
+
+		{
+			description: 'self::node() context, parent::node()/* collapsed like ./../*',
+
+			contextNodeset: 'self::node()',
+			expression: 'parent::node()/bar',
+			expected: ['../bar'],
+		},
+
+		{
+			description: 'self::* context, parent::*/* collapsed like ./../*',
+
+			contextNodeset: 'self::*',
+			expression: 'parent::*/bar',
+			expected: ['../bar'],
+		},
+
+		{
+			description: '// special cases',
+
+			contextNodeset: null,
+			expression: '//. | //./foo | //.. | //../foo',
+			expected: ['//.', '//foo', '//..'],
```

Comment:
Why isn't the last one not `//../foo`?

Reply:
It's resolved to `//foo`. The result is set-like, so that result corresponds to both `//./foo` (second sub-expression) and `//../foo` (fourth/last sub-expression) in the UnionExpr. (This surprised me for a second too, when I wrote the test case.)

But now that I think about it a bit more, I suppose there's a slight chance I'm wrong about `//foo`, `//./foo`, `//../foo` all being equivalent. I think the self step case has to be, but there's a bit of nuance in the parent step case: it means "`foo` which has one or more child elements"[^1]. I don't know how much difference it'd make for this change now, as neither would actually resolve anything with our current downstream usage. But I'm happy to reconsider the handling of this case now, or add a comment about this nuance for future reference in case it becomes an issue when downstream can surface it.

[^1]: This kind of thing is why I find XPath expressivity both impressive and frustrating. It's kind of like regex that way.

Reply:
Ahh, yes, I understand. How about making it `//../bar` so it's a distinct case? I didn't notice there were only three items in the expected set. I got tripped up by that in another place as well.

> "foo which has one or more child elements"

Parent, right? But yeah, I don't think it needs to be addressed.

Reply:
> Ahh, yes, I understand. How about making it //../bar so it's a distinct case? I didn't notice there were only three items in the expected set. I got tripped up by that in another place as well.

Despite both of us being tripped up on the same thing, I think it's valuable to have tests specifically around the set-like behavior. My instinct is:

- where there are cases that exhibit the set-like behavior, break them up so there is a clearer connection between each individual sub-expression and the resolved path

- add one or more dedicated cases specifically around the set-like behavior, where there are _only_ multiple sub-expressions which would resolve to a single path

Do you think that would be clearer for your understanding too?

> Parent, right? But yeah, I don't think it needs to be addressed.

Neither actually! I started to walk through the steps to explain my reasoning, but that helped me realize it _is equivalent_ to `//foo` (probably retreading similar reasoning when I originally arrived at the same result):

`//../foo` is equivalent to:
`/descendant-or-self::node()/../foo` is equivalent to:

1. All nodes in the document except attributes and namespaces (which are not children/descendants):
    - `/`
    - every element
    - every text node
    - every comment node

2. The subset of 1 which are parents of any other subset of 1:
    - `/`
    - every element with any child node

3. The children of 2 which pass the `foo` NameTest

That's equivalent to `//foo`, because `foo` can't be a child of anything filtered out in step 2.

Reply:
> Do you think that would be clearer for your understanding too?

Yes!

> That's equivalent to //foo, because foo can't be a child of anything filtered out in step 2.

Ah, ok.